File: Workflow-Python/scenarios/scenario_1_run_and_wait.py
"""
Scenario 1: Run Task and Wait for Completion ⭐ Most Common

Business Need:
Start a workflow task, wait for it to complete, and then get the results.

Use Cases:
- Need to synchronously wait for task completion
- Need to get the complete output results of the task
- Suitable for single task execution

Usage Steps:
1. Modify API_KEY and WORKFLOW_ID below
2. Modify input_parameters according to your workflow
3. Run script: python scenarios/scenario_1_run_and_wait.py
"""

import os
import time
import traceback
import requests

# ============ Configuration Area - Please modify the following variables ============
# API Key, get from: https://www.browseract.com/reception/integrations
API_KEY = os.getenv("BROWSERACT_API_KEY", "app-abcdefghijklmn")

# Workflow ID, get from: https://www.browseract.com/reception/workflow-list
WORKFLOW_ID = 1234567890

# Workflow input parameters (modify according to your workflow definition)
INPUT_PARAMETERS = [
    {
        "name": "target_url",
        "value": "https://www.google.com/search?q=iphone17",
    },
    {
        "name": "product_limit",
        "value": "10",
    }
]

# Polling configuration
POLL_INTERVAL = 5  # Check task status every 5 seconds
MAX_WAIT_TIME = 600  # Maximum wait time: 10 minutes (600 seconds)
# ================================================

API_BASE_URL = "https://api.browseract.com/v2/workflow"

def run_task(workflow_id, input_parameters):
    """Start a task"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    data = {
        "workflow_id": workflow_id,
        "input_parameters": input_parameters,
        "save_browser_data": True,
    }
    
    api_url = f"{API_BASE_URL}/run-task"
    response = requests.post(api_url, json=data, headers=headers)
    
    if response.status_code == 200:
        result = response.json()
        task_id = result["id"]
        print(f"✅ Task started, Task ID: {task_id}")
        if "profileId" in result:
            print(f"   Profile ID: {result['profileId']}")
        return task_id
    else:
        print(f"❌ Failed to start task: {response.json()}")
        return None

def get_task_status(task_id):
    """Get task status"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    api_url = f"{API_BASE_URL}/get-task-status?task_id={task_id}"
    response = requests.get(api_url, headers=headers)
    
    if response.status_code == 200:
        return response.json().get("status")
    else:
        print(f"⚠️ Failed to get task status: {response.json()}")
        return None

def get_task(task_id):
    """Get detailed task information"""
    headers = {
        "Authorization": f"Bearer {API_KEY}"
    }
    
    api_url = f"{API_BASE_URL}/get-task?task_id={task_id}"
    response = requests.get(api_url, headers=headers)
    
    if response.status_code == 200:
        return response.json()
    else:
        print(f"⚠️ Failed to get task details: {response.json()}")
        return None

def wait_for_task_completion(task_id):
    """Wait for task completion"""
    start_time = time.time()
    
    print(f"\n⏳ Waiting for task completion (max wait time: {MAX_WAIT_TIME} seconds)...")
    
    while True:
        # Check if timeout
        elapsed_time = time.time() - start_time
        if elapsed_time > MAX_WAIT_TIME:
            print(f"\n⏰ Wait timeout (waited {elapsed_time:.0f} seconds)")
            return None
        
        # Get task status
        status = get_task_status(task_id)
        
        if status is None:
            print("⚠️ Unable to get task status, continuing to wait...")
        elif status == "finished":
            print(f"\n✅ Task completed!")
            return "finished"
        elif status == "failed":
            print(f"\n❌ Task execution failed")
            return "failed"
        elif status == "canceled":
            print(f"\n🚫 Task canceled")
            return "canceled"
        else:
            # running, created, paused, etc.
            elapsed = int(elapsed_time)
            print(f"   Status: {status} (waited {elapsed} seconds)", end="\r")
        
        # Wait before checking again
        time.sleep(POLL_INTERVAL)

def main():
    print("=" * 60)
    print("Scenario 1: Run Task and Wait for Completion")
    print("=" * 60)
    
    try:
        # Step 1: Start task
        print("\n📤 Step 1: Starting task...")
        task_id = run_task(WORKFLOW_ID, INPUT_PARAMETERS)
        
        if task_id is None:
            print("❌ Unable to start task, exiting")
            return
        
        # Step 2: Wait for task completion
        print("\n📥 Step 2: Waiting for task completion...")
        final_status = wait_for_task_completion(task_id)
        
        # Step 3: Get task results
        print("\n📊 Step 3: Getting task results...")
        task_info = get_task(task_id)
        
        if task_info:
            print("\n" + "=" * 60)
            print("Task Details:")
            print("=" * 60)
            print(f"Task ID: {task_info.get('id')}")
            print(f"Status: {task_info.get('status')}")
            print(f"Created At: {task_info.get('created_at')}")
            print(f"Finished At: {task_info.get('finished_at')}")
            
            # Display output
            output = task_info.get('output', {})
            if output.get('string'):
                print(f"\nOutput Text:")
                print(output['string'])
            if output.get('files'):
                print(f"\nOutput Files:")
                for file_url in output['files']:
                    print(f"  - {file_url}")
            
            # Display step information
            steps = task_info.get('steps', [])
            if steps:
                print(f"\nExecution Steps ({len(steps)} steps):")
                for step in steps:
                    status_icon = "✅" if step.get('status') == 'succeed' else "❌"
                    print(f"  {status_icon} Step {step.get('step')}: {step.get('step_goal', '')[:50]}...")
            
            # Display failure information
            failure_info = task_info.get('task_failure_info')
            if failure_info:
                print(f"\n❌ Failure Reason:")
                print(f"   {failure_info.get('message', '')}")
            
            print("=" * 60)
        else:
            print("⚠️ Unable to get task details")
            
    except Exception as e:
        error = traceback.format_exc()
        print(f"\n❌ Error occurred:\n{error}")

if __name__ == "__main__":
    main()

File: Scenarios-NodeJs/scenario1_runAndWait.js
/**
 * Scenario 1: Run Task and Wait for Completion ⭐ Most Common
 * 
 * Business Need:
 * Start a workflow task, wait for it to complete, and then get the results.
 * 
 * Use Cases:
 * - Need to synchronously wait for task completion
 * - Need to get the complete output results of the task
 * - Suitable for single task execution
 * 
 * Usage Steps:
 * 1. Modify API_KEY and WORKFLOW_ID below
 * 2. Modify input_parameters according to your workflow
 * 3. Run: node scenario1_runAndWait.js
 */

const https = require('https');

// ============ Configuration Area - Please modify the following variables ============
// API Key, get from: https://www.browseract.com/reception/integrations
const API_KEY = process.env.BROWSERACT_API_KEY || "app-abcdefghijklmn";

// Workflow ID, get from: https://www.browseract.com/reception/workflow-list
const WORKFLOW_ID = 1234567890;

// Polling configuration
const POLL_INTERVAL = 5;  // Check task status every 5 seconds
const MAX_WAIT_TIME = 1800;  // Maximum wait time: 30 minutes (1800 seconds)
// ================================================

const API_BASE_URL = "https://api.browseract.com/v2/workflow";

function makeRequest(options, data) {
    return new Promise((resolve, reject) => {
        const req = https.request(options, (res) => {
            let responseData = '';
            
            res.on('data', (chunk) => {
                responseData += chunk;
            });
            
            res.on('end', () => {
                if (res.statusCode === 200) {
                    resolve(JSON.parse(responseData));
                } else {
                    reject(new Error(`HTTP ${res.statusCode}: ${responseData}`));
                }
            });
        });
        
        req.on('error', (error) => {
            reject(error);
        });
        
        req.setTimeout(30000, () => {
            req.destroy();
            reject(new Error('Request timeout'));
        });
        
        if (data) {
            req.write(data);
        }
        req.end();
    });
}

function runTask(workflowId, inputParameters) {
    return new Promise((resolve, reject) => {
        const data = JSON.stringify({
            workflow_id: workflowId,
            input_parameters: inputParameters,
            save_browser_data: true
        });
        
        const options = {
            hostname: 'api.browseract.com',
            port: 443,
            path: '/v2/workflow/run-task',
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${API_KEY}`,
                'Content-Type': 'application/json',
                'Content-Length': Buffer.byteLength(data)
            }
        };
        
        makeRequest(options, data)
            .then(result => {
                const taskId = result.id;
                console.log(`✅ Task started, Task ID: ${taskId}`);
                if (result.profileId) {
                    console.log(`   Profile ID: ${result.profileId}`);
                }
                resolve(taskId);
            })
            .catch(error => {
                console.log(`❌ Failed to start task: ${error.message}`);
                resolve(null);
            });
    });
}

function getTaskStatus(taskId) {
    return new Promise((resolve) => {
        const options = {
            hostname: 'api.browseract.com',
            port: 443,
            path: `/v2/workflow/get-task-status?task_id=${taskId}`,
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${API_KEY}`
            }
        };
        
        makeRequest(options)
            .then(result => {
                resolve(result.status);
            })
            .catch(() => {
                // Network error, will retry in next polling cycle
                resolve(null);
            });
    });
}

function getTask(taskId) {
    return new Promise((resolve) => {
        const options = {
            hostname: 'api.browseract.com',
            port: 443,
            path: `/v2/workflow/get-task?task_id=${taskId}`,
            method: 'GET',
            headers: {
                'Authorization': `Bearer ${API_KEY}`
            }
        };
        
        makeRequest(options)
            .then(result => {
                resolve(result);
            })
            .catch(error => {
                console.log(`⚠️ Network error while getting task details: ${error.message}`);
                resolve(null);
            });
    });
}

async function waitForTaskCompletion(taskId) {
    const startTime = Date.now();
    let previousStatus = null;
    
    console.log(`\n⏳ Waiting for task completion (max wait time: ${MAX_WAIT_TIME / 60} minutes)...`);
    
    while (true) {
        // Check if timeout
        const elapsedTime = Math.floor((Date.now() - startTime) / 1000);
        if (elapsedTime > MAX_WAIT_TIME) {
            console.log(`\n⏰ Wait timeout (waited ${elapsedTime} seconds)`);
            return null;
        }
        
        // Get task status
        const status = await getTaskStatus(taskId);
        
        if (status === null) {
            // Network error or API error, continue waiting
            console.log();
            process.stdout.write(`   Network error, retrying... (waited ${elapsedTime} seconds)\r`);
        } else if (status === "finished") {
            console.log(`\n✅ Task completed!`);
            return "finished";
        } else if (status === "failed") {
            console.log(`\n❌ Task execution failed`);
            return "failed";
        } else if (status === "canceled") {
            console.log(`\n🚫 Task canceled`);
            return "canceled";
        } else {
            // running, created, paused, etc.
            // If status changed, print on new line; otherwise update same line
            if (status !== previousStatus) {
                console.log();  // New line when status changes
                process.stdout.write(`   Status: ${status} (waited ${elapsedTime} seconds)\r`);
                previousStatus = status;
            } else {
                // Same status, update same line
                process.stdout.write(`   Status: ${status} (waited ${elapsedTime} seconds)\r`);
            }
        }
        
        // Wait before checking again
        await new Promise(resolve => setTimeout(resolve, POLL_INTERVAL * 1000));
    }
}

async function main() {
    console.log("============================================================");
    console.log("Scenario 1: Run Task and Wait for Completion");
    console.log("============================================================");
    
    try {
        // Step 1: Start task
        console.log("\n📤 Step 1: Starting task...");
        
        // Workflow input parameters (modify according to your workflow definition)
        const inputParameters = [
            {
                name: "target_url",
                value: "https://www.google.com/search?q=iphone17"
            },
            {
                name: "product_limit",
                value: "10"
            }
        ];
        
        const taskId = await runTask(WORKFLOW_ID, inputParameters);
        
        if (!taskId) {
            console.log("❌ Unable to start task, exiting");
            return;
        }
        
        // Step 2: Wait for task completion
        console.log("\n📥 Step 2: Waiting for task completion...");
        const finalStatus = await waitForTaskCompletion(taskId);
        
        // Step 3: Get task results
        console.log("\n📊 Step 3: Getting task results...");
        const taskInfo = await getTask(taskId);
        
        if (taskInfo) {
            console.log("\n============================================================");
            console.log("Task Result (JSON):");
            console.log("============================================================");
            // Output complete task result as formatted JSON
            console.log(JSON.stringify(taskInfo, null, 2));
            console.log("============================================================");
        } else {
            console.log("⚠️ Unable to get task details");
        }
        
    } catch (error) {
        console.error("\n❌ Error occurred:");
        console.error(error);
    }
}

main();
